### PR TITLE
Enables AMR to be used in GH executables with one or two black holes

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

This adds the code that will hook AMR into the single/binary black hole executables.
Currently only will work for p-refinement with global time-stepping (otherwise will runtime error when an AMR change is triggered).  The example GH input files do not trigger refinement.

### Upgrade instructions

Input files for these executables will need to specify
```
Amr:
  Criteria:
  Verbosity: Quiet
```
in order to run, but do no AMR.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments
This should be considered as experimental support not yet ready for adoption.  My plan is to work on the following, the first two to be done ASAP so others can more easily experiment:
- Allow AMR policies (such as isotropic vs anisotropic refinement) to be set via the input file
- An AMR tutorial with documentation on how to enable and control AMR, and how to add it to other evolution systems
- enable support for local time-stepping with p-refinement
- enable support for h-refinement

### CODE REVIEW

File: src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp [hunk 1]
```diff
@@ -153,7 +153,11 @@ class Constraints : public Criterion {
   WRAPPED_PUPable_decl_template(Constraints);  // NOLINT
   /// \endcond
 
-  using compute_tags_for_observation_box = TensorTags;
+  using compute_tags_for_observation_box = tmpl::append<
+      TensorTags, tmpl::list<domain::Tags::JacobianCompute<
+                                 Dim, Frame::ElementLogical, Frame::Inertial>,
+                             Events::Tags::ObserverJacobianCompute<
```
Comment (knelli2) [inline src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp:159]: Why both?

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveGhSingleBlackHole.hpp [hunk 2]
```diff
@@ -98,7 +98,7 @@ struct EvolutionMetavars : public GeneralizedHarmonicTemplateBase<3> {
         tmpl::list<intrp::callbacks::FindApparentHorizon<ApparentHorizon,
                                                          ::Frame::Inertial>>;
     using horizon_find_failure_callback =
-        intrp::callbacks::IgnoreFailedApparentHorizon;
+        intrp::callbacks::ErrorOnFailedApparentHorizon;
```
Comment (knelli2) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveGhSingleBlackHole.hpp]: Why change this? The reason it was ignore was because if something weird happens with the observation horizon finder, we didn't want the evolution to be affected.

File: src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.hpp [hunk 3]
```diff
@@ -207,8 +207,9 @@ struct ProjectMortars : tt::ConformsTo<amr::protocols::Projector> {
       const Mesh<dim>& new_mesh, const Element<dim>& new_element,
       const std::unordered_map<ElementId<dim>, amr::Info<dim>>& neighbor_info,
       const std::pair<Mesh<dim>, Element<dim>>& /*old_mesh_and_element*/) {
-    static_assert(not Metavariables::local_time_stepping,
-                  "AMR with local time-stepping is not yet supported");
+    if (Metavariables::local_time_stepping) {
+      ERROR("AMR with local time-stepping is not yet supported");
```
Comment (knelli2) [inline src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.hpp:211]: Do you know if there's an easy way to have this error happen at option parsing? If not, don't worry about it.

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveGhSingleBlackHole.hpp [hunk 4]
```diff
@@ -239,6 +269,8 @@ struct EvolutionMetavars : public GeneralizedHarmonicTemplateBase<3> {
                          Actions::AdvanceTime,
                          PhaseControl::Actions::ExecutePhaseChange>>>>>;
 
+  using dg_element_array = gh_dg_element_array;
+
```
Comment (knelli2) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveGhSingleBlackHole.hpp]: Why is this necessary?

**Review by knelli2 — COMMENTED** [review]
Just a few questions

File: src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp (hunk 1 above)
Comment (kidder) [reply src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp:159]: the second one depends upon the first

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveGhSingleBlackHole.hpp (hunk 2 above)
Comment (kidder) [reply src/Evolution/Executables/GeneralizedHarmonic/EvolveGhSingleBlackHole.hpp]: oops, leftover from when I wanted the evolution to stop when it could no longer find the horizon

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveGhSingleBlackHole.hpp (hunk 4 above)
Comment (kidder) [reply src/Evolution/Executables/GeneralizedHarmonic/EvolveGhSingleBlackHole.hpp]: because `amr::Component<Metavariables>` initiates AMR on `Metavariables::dg_element_array`

Comment (knelli2) [reply src/Evolution/Executables/GeneralizedHarmonic/EvolveGhSingleBlackHole.hpp]: Would it be feasible to put this type alias inside the `amr` protocol struct in the metavars? That way it's documented and isn't another magic type alias.

Comment (nilsvu) [reply src/Evolution/Executables/GeneralizedHarmonic/EvolveGhSingleBlackHole.hpp]: I like that 👍 Or pass it down as a template param.

File: src/ParallelAlgorithms/Amr/Projectors/CopyFromCreatorOrLeaveAsIs.hpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                        children_items) {
+    const auto& first_child_items = children_items.begin();
+    for (auto it = std::next(first_child_items); it != children_items.end();
+         ++it) {
+      const bool children_agree =
+          ((tuples::get<Tags>(first_child_items->second) ==
+            tuples::get<Tags>(it->second)) and
+           ...);
+      ASSERT(children_agree, "Children do not agree on all items!");
+    }
```
Comment (nilsvu) [inline src/ParallelAlgorithms/Amr/Projectors/CopyFromCreatorOrLeaveAsIs.hpp:72]: Wrap in `#ifdef SPECTRE_DEBUG`.

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveGhSingleBlackHole.hpp [hunk 6]
```diff
@@ -187,6 +187,33 @@ struct EvolutionMetavars : public GeneralizedHarmonicTemplateBase<3> {
...
+            Tags::StepperError<typename system::variables_tag>,
+            Tags::PreviousStepperError<typename system::variables_tag>,
+            Tags::StepperErrorUpdated,
+            SelfStart::Tags::InitialValue<typename system::variables_tag>,
+            SelfStart::Tags::InitialValue<Tags::TimeStep>,
+            SelfStart::Tags::InitialValue<Tags::Next<Tags::TimeStep>>>,
+        ::amr::projectors::CopyFromCreatorOrLeaveAsIs<tmpl::push_back<
+            typename control_system::Actions::InitializeMeasurements<
+                control_systems>::simple_tags,
+            intrp::Tags::InterpPointInfo<EvolutionMetavars>>>>;
```
Comment (nilsvu) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveGhSingleBlackHole.hpp]: Can you move more of these into projectors so the decision what to do with these tags doesn't have to be made in the metavars? The module that adds these tags know what to do with them, the user adding AMR support to the metavars probably doesn't.

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp [hunk 7]
```diff
@@ -484,19 +510,23 @@ struct EvolutionMetavars {
                  gh::ConstraintDamping::Tags::DampingFunctionGamma1<
                      volume_dim, Frame::Grid>,
                  gh::ConstraintDamping::Tags::DampingFunctionGamma2<
-                     volume_dim, Frame::Grid>>;
+                     volume_dim, Frame::Grid>,
+                 ::amr::Criteria::Tags::Criteria>;
```
Comment (nilsvu) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp]: not needed anymore

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveGhSingleBlackHole.hpp (hunk 6 above)
Comment (kidder) [reply src/Evolution/Executables/GeneralizedHarmonic/EvolveGhSingleBlackHole.hpp]: I agree that this needs changing, but that is beyond the scope of this PR.  

- Currently multiple actions are allowed to add the same tag so there is no unique module that adds them (and furthermore some of these tags are added my non-initialization actions)
- DefaultInitialize and CopyFromCreatorOrLeaveAsIs are generic projectors that would be copied multiple times in specific projectors
- I would like to get LTS working before tackling these issues.

**Review by knelli2 — COMMENTED** [review]
No additional comments besides clang-tidy

**Comment by kidder** [comment]
@nilsvu please look at the fixups

**Review by nilsvu — DISMISSED** [review]

**Review by nilsvu — APPROVED** [review]

**Review by knelli2 — APPROVED** [review]
